Add serialisation support to SeqAn3

h-2 opened the pull request:
This adds proper serialisation support. It is completely optional so SeqAn will build with and without Cereal. You don't have to do macro voodoo when adding serialisation functions as the serialisation functions are templates.

Have a look at f13d7db how easy this for specific types.

Generic versions are more difficult, but we won't need to implement them often.

This PR also adds `-pedantic` to our tests, because I was having warnings and fixing them anyway.

Please also note the changes here:
* https://github.com/seqan/seqan3/wiki/File-structure-and-naming#contents (order of includes, optional deps go **after SeqAn**)
* https://github.com/seqan/seqan3/wiki/Documentation#configuration (some changes to doxygen config)

h-2 commented on include/seqan3/core/concept/cereal.hpp:
```diff
@@ -0,0 +1,153 @@
+{
+
+template <typename type>
+using strip_cereal_wrapper_t = typename cereal::traits::strip_minimal<std::decay_t<type>>::type;
```
The stuff in this file could all be done with a `static constexpr bool seqan_with_serial` that's no problem. But in other cases it's not possible.

h-2 replied:
>so the second code path is likely not to be as tested as the other one.
> Thus it might happen that we add something to the first code path and will forgot to update the second one.

It definitely needs the same amount of testing.

> You could also make local #ifdefs like

Yeah, I could re-order it, but it increases the macro verbosity.

h-2 commented on include/seqan3/core/concept/cereal.hpp:
```diff
@@ -0,0 +1,153 @@
+ *
+ * \attention
+ * The cereal library is an optional dependency of SeqAn, if it is not found **no types** satisfy this concept.
+ */
```
Well there are two definitions, one with cereal, one without. I picked the one that is parsed by doxygen.

h-2 replied:
done

h-2 commented on include/seqan3/core/platform.hpp:
```diff
@@ -90,7 +90,39 @@
+#   else
+#       define SEQAN_WITH_CEREAL 0
+#   endif
+#endif
```
> I don't like macros. Can't we do it without them?

I don't like them either, but we can't do without them. Have a look at the tests for example-

> If we really use them, they should be called SEQAN3_*

Hm, yeah, we could discuss this.

> Why not defining a constexpr bool seqan3::cereal_support

We can discuss adding this *additionally* and replacing the code in core/concept/cereal.hpp if you insist :smile: 

> As later parts indicate, we require that the user first includes cereal/cereal.hpp before including our library. Is that right?

No, it's not required. In fact I recommend the opposite now.

> But __has_include does not ensure that the library was already included.

`__has_include` only checks for the existance of a file, it doesn't check whether it was loaded.

h-2 replied:
ok

h-2 commented on include/seqan3/core/platform.hpp:
```diff
@@ -90,7 +90,39 @@
+    /*!\cond DEV
+     * \name Cereal function macros
+     * \ingroup core
+     * \brief These can be changed by apps so we used the macros instead of the values internally.
```
This is part of CEREAL's design. It allows you to incorporate cereal in your library and avoid naming conflicts in case any of the mentioned function names is already used for something else. We just make sure that if there ever is someone who uses this feature, it will work with seqan, too. 
Our documentation system automatically replaces these macros with simple function names, though, so it's no problem for our users (they don't use these functions directly anyway).

h-2 commented on include/seqan3/core/platform.hpp:
```diff
@@ -90,7 +90,39 @@
+#   define CEREAL_LOAD_FUNCTION_NAME load
+#   define CEREAL_SAVE_FUNCTION_NAME save
+#   define CEREAL_LOAD_MINIMAL_FUNCTION_NAME load_minimal
+#   define CEREAL_SAVE_MINIMAL_FUNCTION_NAME save_minimal
```
no, no problem

h-2 replied:
> Of course it's a problem:

Only if they are building **without cereal**, in which case the MACROs are useless anyway and modifying them makes no sense and probably should create an error.

h-2 commented on include/seqan3/range/container/concatenated_sequences.hpp:
```diff
@@ -49,11 +49,16 @@
 
+#if SEQAN_WITH_CEREAL
+#include <cereal/types/vector.hpp>
+#endif
```
> I don't see that you use this include.

You need it, when you serialise concatenated_sequences, because it contains a vector as member (likely even two) which are serialised.

> Isn't that included in cereal/cereal.hpp?

No. Support for serialising STL types needs to be explicitly loaded. One could not include it here and demand the user that wants to serialise add this, but how should a user know that internally concatenated_sequences uses a vector?

> I would use if __has_include(<cereal/types/vector.hpp>)

No, this would be bad, because it would make the compiler search the entire include directories every time and also we want to be able to deactivate cereal, even if it is there – for testing, but not only. That's another reason we have a macro.

h-2 commented on include/seqan3/range/container/concatenated_sequences.hpp:
```diff
@@ -1282,6 +1288,20 @@ class concatenated_sequences
+     * \attention These functions are never called directly, see \ref serialisation for more details.
+     */
+    template <cereal_archive_concept archive_t>
+    void CEREAL_SERIALIZE_FUNCTION_NAME(archive_t & archive)
```
No, every header file is required to include platform so that setups without cereal get the fallback definitions. Setups with cereal get `<cereal/cereal.hpp> ` by including `<seqan3/core/concept/cereal.hpp>` – which they need to include anyway for `cereal_archive_concept`.

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -33,10 +33,20 @@
 
 #include <seqan3/alphabet/all.hpp>
 
+#if SEQAN_WITH_CEREAL
```
> again check directly with #if __has_include() it's more easy to understand than some self written macros.

1) see above, we need to be able to deactivate, even if its there
2) it would be much harder to maintain. Imagine they rename the top-leve file, then we have to search through all of our code-base to change this. Right now we need to change the macro and a single include in core.
3) Also imagine there is cereal on disk, but it's the wrong version. A macro can have multiple conditions, adding multiple conditions to every file will get really complicated.

> And this are our test cases so it makes no sense that a user may disable them. For the library itself it might be sensible to be able to disable it, but the test cases are our domain.

But the test cases are supposed to ensure it works for our users → we have to test if our library works without cereal, because our users might not have it!!

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -214,6 +224,55 @@ TYPED_TEST(alphabet, concept)
+        std::ofstream os(filename, std::ios::binary);
+        out_archive_t archive(os);
+        archive(l);
+        archive(vec);
```
This is cereal's interface and I can't change it. It's not a global function, but the `() operator` of the actual archive. I am not convinced it's the best solution, but it's better than other things I have seen.

h-2 replied:
ah, ok, yeah that's possible, I was just against naming it something that resembles a function name. Also maybe we should use `{}` in construction to differentiate more clearly from the `operator()`.

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -214,6 +224,55 @@ TYPED_TEST(alphabet, concept)
+        EXPECT_EQ(vec, in_vec);
+    }
+
+    std::remove(filename.c_str());
```
You are right :bulb: There are other problems with the current approach, but I didn't immediately find a nicer solution, because the safer solutions delete the file when it is closed, but our file needs to be opened and closed multiple times to simulate proper serialisation and de-serialisation.

marehr replied:
We could put that into a `unique_ptr` with a custom deleter and have an utility function that creates that for our testcases?

```cpp
std::unique_ptr<char, delctype(&std::remove)> filename(std::tmpnam(nullptr), &std::remove);
```

In http://en.cppreference.com/w/cpp/memory/unique_ptr there is also this example:
```cpp
  {
      std::unique_ptr<std::FILE, decltype(&std::fclose)> fp(std::fopen("demo.txt", "r"),
                                                            &std::fclose);
      if(fp) // fopen could have failed; in which case fp holds a null pointer
        std::cout << (char)std::fgetc(fp.get()) << '\n';
  } // fclose() called here, but only if FILE* is not a null pointer
```
Maybe we should http://en.cppreference.com/w/cpp/io/c/tmpfile instead.

h-2 replied:
> We could put that into a unique_ptr with a custom deleter and have a utility function that creates that for our testcases?

Yes, something like that would be possible, I will check it out.

> Maybe we should use http://en.cppreference.com/w/cpp/io/c/tmpfile instead.

No, this won't work, because, as I said above:

> > the safer solutions delete the file when it is closed, but our file needs to be opened and closed multiple times to simulate proper serialisation and de-serialisation.

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -214,6 +224,55 @@ TYPED_TEST(alphabet, concept)
+    if constexpr (alphabet_size_v<TypeParam>)
+        assign_rank(letter, 1);
+    else
+        assign_rank(letter, 1);
```
Lol, you are right, it should be `assign_rank(letter, 0);` in the second branch!

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -214,6 +224,55 @@ TYPED_TEST(alphabet, concept)
+
+    std::vector<TypeParam> vec;
+    vec.resize(10);
+    if constexpr (alphabet_size_v<TypeParam>)
```
if you insist

h-2 replied:
Ah, I messed up, it should actually be > 1. You just convinced me that it is better to write out the condition :laughing:

h-2 commented on test/alphabet/alphabet_test.cpp:
```diff
@@ -214,6 +224,55 @@ TYPED_TEST(alphabet, concept)
+    vec.resize(10);
+    if constexpr (alphabet_size_v<TypeParam>)
+        for (unsigned i = 0; i < 10; ++i)
+            assign_rank(vec[i], i % 2);
```
yep, see above

h-2 commented on test/range/container/container_of_container_test.cpp:
```diff
@@ -33,14 +33,24 @@
 #include <seqan3/range/container/all.hpp>
 #include <seqan3/range/view/to_char.hpp>
 
+#if SEQAN_WITH_CEREAL
```
no, see above

h-2 commented on test/range/container/container_of_container_test.cpp:
```diff
@@ -345,3 +355,39 @@ TYPED_TEST(container_of_container, swap)
     EXPECT_TRUE((t1 == TypeParam{}));
 }
+
+#if SEQAN_WITH_CEREAL
```
no, see above

h-2 commented on test/range/container/container_of_container_test.cpp:
```diff
@@ -345,3 +355,39 @@ TYPED_TEST(container_of_container, swap)
+
+    {
+        std::ofstream os(filename, std::ios::binary);
+        out_archive_t archive(os);
```
can't, see above

marehr replied:
Simply rename `archive` into `oarchive`

h-2 commented on test/CMakeLists.txt:
```diff
@@ -65,7 +65,16 @@ include_directories("${GOOGLETEST_ROOT}/googletest/include/")
+
+if (NO_CEREAL)
+    add_definitions (-DSEQAN_WITH_CEREAL=0)
+endif (NO_CEREAL)
```
we need to actually test the codepaths our users can see, i.e. with and without cereal; see above

h-2 replied:
The thing is that we need to test all test cases that include files that have serialisation code, not only the test cases that actually use this code (because of include and definition errors). And I suspect more than 90% of our test cases include some files that include some serialisation code so it really affects everything.
Just think about it: anything that uses an alphabet, anything that uses a custom container like `seqan3::concatenated_sequences`, all future data structures like alignments or indexes...

We can think about not making every CI run execute everything, but for the nightlies I want all combinations and for CI I want at least one run without CEREAL just to be sure. Since we only test gcc7 right now, I suggest we just have one run CEREAL and one without.

h-2 commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -124,4 +125,54 @@ concept bool alphabet_concept = requires (t t1, t t2)
+ *
+ * Delegates to seqan3::semi_alphabet_concept::to_rank.
+ *
+ * \attention These functions are never called directly, see \ref alphabet on how to use serialisation.
```
It's the alphabet module page! I know it's super inconsistent when links are automatically generated and how you manually do it, if they aren't auto-generated.

h-2 replied:
👍

h-2 commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -124,4 +125,54 @@ concept bool alphabet_concept = requires (t t1, t t2)
+ * \attention These functions are never called directly, see \ref alphabet on how to use serialisation.
+ */
+template <cereal_output_archive_concept archive_t, semi_alphabet_concept alphabet_t>
+underlying_rank_t<alphabet_t> CEREAL_SAVE_MINIMAL_FUNCTION_NAME(archive_t const &, alphabet_t const & l)
```
There three different serialisation interfaces, I can explain it to you AFK, if you want to!

h-2 commented:
@marehr Thanks for the complete review. We can discuss this in more detail tomorrow!

h-2 commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -124,4 +125,54 @@ concept bool alphabet_concept = requires (t t1, t t2)
+                                       underlying_rank_t<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>> const & r)
+    requires semi_alphabet_concept<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>>
+{
+    assign_rank(static_cast<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>&&>(l), r);
```
> These functions (save_minimal and load_minimal) does not need to be in the namespace cereal?

no

> Your code seems overly complicated. With all that detail::strip_cereal_wrapper_t stuff.

Yeah, well I didn't want to do it that way 😥 
See USCiLab/cereal/issues/436 for the reasons that I had to do it this way.

> I needed serialization for what I'm working on and I simply wrote the following:

Yeah, but we really want the `_minimal` versions here!

h-2 commented:
Why do some of your review comments appear twice 🤔 
Anyway, everything should be answered. I will incorporate your suggestions tomorrow or Friday.

marehr commented:
Okay maybe you can explain me the differences when you are back 😉

Am 1. November 2017 18:04:16 MEZ schrieb Hannes Hauswedell <notifications@github.com>:
>h-2 commented on this pull request.
>
>
>
>> + * \param r The assigned value.
>+ * \relates seqan3::semi_alphabet_concept
>+ *
>+ * \details
>+ *
>+ * Delegates to seqan3::semi_alphabet_concept::assign_rank.
>+ *
>+ * \attention These functions are never called directly, see \ref
>alphabet on how to use serialisation.
>+ */
>+template <cereal_input_archive_concept archive_t, typename
>wrapped_alphabet_t>
>+void CEREAL_LOAD_MINIMAL_FUNCTION_NAME(archive_t const &,
>+                                       wrapped_alphabet_t && l,
>+                                      
>underlying_rank_t<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>>
>const & r)
>+    requires
>semi_alphabet_concept<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>>
>+{
>+   
>assign_rank(static_cast<detail::strip_cereal_wrapper_t<wrapped_alphabet_t>&&>(l),
>r);
>
>> These functions (save_minimal and load_minimal) does not need to be
>in the namespace cereal?
>
>no
>
>> Your code seems overly complicated. With all that
>detail::strip_cereal_wrapper_t stuff.
>
>Yeah, well I didn't want to do it that way 😥 
>See USCiLab/cereal/issues/436 for the reasons that I had to do it this
>way.
>
>> I needed serialization for what I'm working on and I simply wrote the
>following:
>
>Yeah, but we really want the `_minimal` versions here!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/seqan/seqan3/pull/110#discussion_r148320665

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

h-2 commented:
@marehr I think I addressed all your issues that remained.

@joergi-w I have included a small fix of your tests that broke with new warnings

@sarahet I have included a small include error in the translation module that somehow was not triggered before.
